File: mods/Simulated-Project/simulated/common/src/main/java/dev/simulated_team/simulated/content/blocks/void_anchor/VoidAnchorRenderer.java
package dev.simulated_team.simulated.content.blocks.void_anchor;

import com.mojang.blaze3d.vertex.PoseStack;
import dev.simulated_team.simulated.content.end_sea.EndSeaShadowRenderer;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer;
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider;
import net.minecraft.world.phys.Vec3;

public class VoidAnchorRenderer implements BlockEntityRenderer<VoidAnchorBlockEntity> {
    public VoidAnchorRenderer(final BlockEntityRendererProvider.Context context) {

    }

    @Override
    public boolean shouldRender(final VoidAnchorBlockEntity blockEntity, final Vec3 vec3) {
        return true;
    }

    @Override
    public int getViewDistance() {
        return 512;
    }

    @Override
    public boolean shouldRenderOffScreen(final VoidAnchorBlockEntity blockEntity) {
        return true;
    }

    @Override
    public void render(final VoidAnchorBlockEntity blockEntity, final float f, final PoseStack poseStack, final MultiBufferSource multiBufferSource, final int i, final int j) {
        if (EndSeaShadowRenderer.renderingShadowMap())
            return;

        EndSeaShadowRenderer.addVoidAnchor(blockEntity);
    }
}

File: src/main/kotlin/dev/codinglabs/modpack/client/VoidAnchorRenderer.kt
package dev.codinglabs.modpack.client

import com.mojang.blaze3d.vertex.PoseStack
import dev.codinglabs.modpack.VoidAnchorBlockEntity
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider

class VoidAnchorRenderer(@Suppress("UNUSED_PARAMETER") context: BlockEntityRendererProvider.Context) : BlockEntityRenderer<VoidAnchorBlockEntity> {

    override fun render(
        blockEntity: VoidAnchorBlockEntity,
        partialTick: Float,
        poseStack: PoseStack,
        bufferSource: MultiBufferSource,
        packedLight: Int,
        packedOverlay: Int,
    ) {

        val consumer = bufferSource.getBuffer(RenderType.endPortal())
        val pose = poseStack.last().pose()

        val min = 0.0f
        val max = 1.0f
        val y = 1.002f

        // Top face
        consumer.addVertex(pose, min, y, min)
        consumer.addVertex(pose, min, y, max)
        consumer.addVertex(pose, max, y, max)
        consumer.addVertex(pose, max, y, min)

        // Back face so it is still visible from lower camera angles.
        consumer.addVertex(pose, max, y, min)
        consumer.addVertex(pose, max, y, max)
        consumer.addVertex(pose, min, y, max)
        consumer.addVertex(pose, min, y, min)
    }
}

